### PR TITLE
FXVPN-10: onboarding flow

### PR DESCRIPTION
**Pull request by mcleinman**
These should match [onboarding mocks from figma](https://www.figma.com/design/s13B7zs27cadZXUyvxobgW/Mozilla-VPN-Extension?node-id=1848-41802&node-type=frame&t=k3T4A0zHBsqtwmbk-0).

Associated PR is up for the string: https://github.com/mozilla-l10n/mozilla-vpn-extension-l10n/pull/3

Updated video: https://github.com/user-attachments/assets/317e48e7-9982-4489-ad19-8651dd5f21a6

The scroll bars in the video on page 2 are due to my browser settings, and won't be seen by most people.

### CODE REVIEW

File: src/shared/property.js [hunk 1]
```diff
@@ -250,3 +250,28 @@ export const propertySum = (left, right, transform) => {
...
+ * @template L
+ * @template M
+ * @template R
+ * @param {IBindable<L>} left - Left Hand Property
+ * @param {IBindable<R>} right - Middle Hand Property
+ * @param {IBindable<R>} right - Right Hand Property
+ * @param {(arg0: L, arg1: M, arg2: R)=>T} transform - Called with the Property Value, must return the transformed value
+ * @returns {ReadOnlyProperty<T>} -
+ */
+export const propertySumTrio = (left, middle, right, transform) => {
```
Comment (strseb) [inline src/shared/property.js:271]: A: First of really appreciate that you looked into that system :D 
However i dont think that scales. 
Consider cherry picking https://github.com/mozilla-extensions/mozilla-vpn-extension/pull/144 - i think we can just rewrite property sum to take N amount of properties :)

File: src/background/extensionController/extensionController.js
Comment (strseb) [inline src/background/extensionController/extensionController.js:1]: Nitpick: 
```
 * ExtensionController manages extension state and
 * provides a method to the popup for disabling and enabling
* the "Firefox VPN".
```
I don't think we should therfore have onboarding here. 
All of the code here looks great, but i'd prefer it moved into `onboarding.js` and we expose that in `main.js` similar to the others :)

File: src/background/extensionController/extensionController.js [hunk 2]
```diff
@@ -81,6 +93,23 @@ export class ExtensionController extends Component {
...
+  }
+
+  nextOnboardingPage() {
+    this.#mCurrentOnboardingPage.set(this.#mCurrentOnboardingPage.value + 1);
+  }
+
+  finishOnboarding() {
+    this.#mCurrentOnboardingPage.set(4);
+    putIntoStorage(
+      4,
```
Comment (strseb) [inline src/background/extensionController/extensionController.js:107]: Consider using constants to denote magic numbers Unless you have seen the figma it's not clear the max onboarding page is "4"

Comment (strseb) [reply src/background/extensionController/extensionController.js:107]: Also what happens if we add a new onboarding page? Users would then be seeing that one no? 
(Not sure if this is wanted or not)

File: src/ui/browserAction/popup.html [hunk 3]
```diff
@@ -34,7 +34,16 @@
       <open-mozilla-vpn-message-screen
         slot="MessageOpenMozillaVPN"
       ></open-mozilla-vpn-message-screen>
-      <!-- TODO: Split Tunnel Message, Onboarding, Update Message -->
+      <onboarding-screen-1
```
Comment (strseb) [inline src/ui/browserAction/popup.html:37]: With this approach i can see why the pagination would be hard to add (agreed) 

WDYT: 
one new `Screen` Element -> "Onboarding" that is shown here if onboarding is not complete. 

The Onboarding screen then would be something like 
```html
<main>
<conditional-view slotName=${this.onboardingCount} >
  <slot=1 onboarding-screen-1>
  <slot=2 onboarding-screen-2>
  <slot=3 onboarding-screen-3>
  <slot=4 onboarding-screen-4>
</conditional-view>
<pagination>
<button-next>

</main>


```

**Review by strseb — CHANGES_REQUESTED** [review]

File: src/ui/browserAction/popup.html (hunk 3 above)
Comment (strseb) [reply src/ui/browserAction/popup.html:37]: (this is also a nonblocking suggestion, if we REALLY dont want the indicators let's not add this )

File: src/background/extensionController/extensionController.js (hunk 2 above)
Comment (mcleinman) [reply src/background/extensionController/extensionController.js:107]: Will update away from magic numbers, thanks.

We do want users to have the possibility of later seeing new onboarding pages - Lesley mentioned this was desired (and sometimes difficult) w/ MAC.

**Comment by mcleinman** [comment]
We got a new screen added into the mocs yesterday, so putting this on draft until I can update the PR.

**Comment by lesleyjanenorton** [comment]
> We got a new screen added into the mocs yesterday, so putting this on draft until I can update the PR.

@mcleinman do you mean the update the settings panel? if not, mind shooting me a link to the new screen? thanks!

File: src/ui/browserAction/popup.html (hunk 3 above)
Comment (lesleyjanenorton) [reply src/ui/browserAction/popup.html:37]: I'm not fixated on the pagination either (MAC doesn't have them) but I do think they're a very nice touch, if not too painful.

**Comment by mcleinman** [comment]
> > We got a new screen added into the mocs yesterday, so putting this on draft until I can update the PR.
> 
> @mcleinman do you mean the update the settings panel? if not, mind shooting me a link to the new screen? thanks!

The settings panel also came with a new onboarding panel: https://www.figma.com/design/s13B7zs27cadZXUyvxobgW/Mozilla-VPN-Extension?node-id=1822-31036&node-type=section&t=zKYBItOJwJmrfOoF-0

File: src/background/vpncontroller/vpncontroller.js [hunk 4]
```diff
@@ -269,28 +269,30 @@ const MOZILLA_VPN_SERVERS_KEY = "mozillaVpnServers";
...
+export async function fromStorage(
   storage = browser.storage.local,
-  key = MOZILLA_VPN_SERVERS_KEY,
+  key,
   defaultValue
 ) {
-  const { mozillaVpnServers } = await storage.get(key);
-  if (typeof mozillaVpnServers === "undefined") {
+  const storageRetrieval = await storage.get(key);
+  const returnValue = storageRetrieval[key];
```
[the remark this answers is not in the document]
Comment (mcleinman) [reply src/background/vpncontroller/vpncontroller.js:281]: I added some more defensive code around here. Good call, thanks!

File: src/background/extensionController/extensionController.js
Comment (mcleinman) [reply src/background/extensionController/extensionController.js:1]: Good call, done

File: src/shared/property.js (hunk 1 above)
Comment (mcleinman) [reply src/shared/property.js:271]: Thanks! I knew it was only a quick band aid fix... Will cherry pick that improved one after it is merged in.

File: src/ui/browserAction/popup.html (hunk 3 above)
Comment (mcleinman) [reply src/ui/browserAction/popup.html:37]: Pagination is now in here - will update the video once I add in the new fourth screen.

**Comment by mcleinman** [comment]
I believe this is ready for review. 

The mocks got updated in the middle of this week with a new fourth onboarding screen. I'm inclined to add that after this is merged - I've started working on it, but there are open questions around copy and images, and I'm not sure this needs to be held up on it. Also open to putting this one back on draft if we think it's better to do it all together.

File: src/background/onboarding.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import { PropertyType } from "./../shared/ipc.js";
+import { fromStorage, putIntoStorage } from "./vpncontroller/vpncontroller.js";
+
+const ONBOARDING_KEY = "mozillaVpnOnboarding";
+const FIRST_PAGE = 1;
+export const NUMBER_OF_ONBOARDING_PAGES = 3;
+const FIRST_UNUSED_PAGE = NUMBER_OF_ONBOARDING_PAGES + 1;
+
+// const log = Logger.logger("RequestHandler");
+let self;
```
Comment (lesleyjanenorton) [inline src/background/onboarding.js:19]: unused I think?

File: src/background/onboarding.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import { property } from "../shared/property.js";
+import { PropertyType } from "./../shared/ipc.js";
+import { fromStorage, putIntoStorage } from "./vpncontroller/vpncontroller.js";
+
+const ONBOARDING_KEY = "mozillaVpnOnboarding";
+const FIRST_PAGE = 1;
+export const NUMBER_OF_ONBOARDING_PAGES = 3;
+const FIRST_UNUSED_PAGE = NUMBER_OF_ONBOARDING_PAGES + 1;
+
+// const log = Logger.logger("RequestHandler");
```
Comment (lesleyjanenorton) [inline src/background/onboarding.js:18]: We can remove this

File: src/background/onboarding.js (hunk 5 above)
Comment (mcleinman) [reply src/background/onboarding.js]: Removed, whoops. Thanks.

File: src/background/onboarding.js (hunk 6 above)
Comment (mcleinman) [reply src/background/onboarding.js]: Removed, whoops. Thanks.

File: src/components/prefab-screens.js [hunk 7]
```diff
@@ -34,10 +39,23 @@ const defineMessageScreen = (
       this.img = img;
       this.heading = heading;
       this.primaryAction = primaryAction;
-      this.onPrimaryAction = onPrimaryAction;
-      this.secondaryAction = secondarAction;
-      this.onSecondaryAction = onSecondaryAction;
+      this.secondaryAction = secondaryAction;
+      if (closeOnClick) {
```
Comment (strseb) [inline src/components/prefab-screens.js:43]: I would like to make this as stupid as it needs to be - WDYT: 
Remove the logic for closeOnClick as it only applys for the actions - and use a decorator for the ones where we need that functionality. 
```js
const closeAfter (f) => { 
          if(f){
            f();
          }
          window.close();
} 
```

```js
const defineMessageScreen = (
   ...
  onSecondaryAction = () => closeAfter (()=>open(sumoLink)),
)
```

Imho that is nicer to read then, as we dont need to recall what that one bool is flagging
```
defineMessageScreen( ... false|true ... )

**Review by strseb — APPROVED** [review]
r+wc, thank you !!

File: src/components/prefab-screens.js (hunk 7 above)
Comment (mcleinman) [reply src/components/prefab-screens.js]: I've updated this. Not my personal style, but feels more appropriate given this project's overall codebase.

**Pull request merged**